src/gui/touchscreenlayout.cpp

```diff
@@ -0,0 +1,348 @@
+		touch_gui_button_id btn = (touch_gui_button_id)i;
+		if (isButtonAllowed(btn) && layout.count(btn) == 0)
+			missing_buttons.push_back(btn);
+	}
+	return missing_buttons;
+}
+
+void ButtonLayout::serializeJson(std::ostream &os) const
+{
+	Json::Value root = Json::objectValue;
+	root["layout"] = Json::objectValue;
+
+	for (const auto &[id, meta] : layout) {
+		Json::Value button = Json::objectValue;
+		button["position_x"] = meta.position.X;
+		button["position_y"] = meta.position.Y;
+		button["offset_x"] = meta.offset.X;
+		button["offset_y"] = meta.offset.Y;
+
+		root["layout"][button_names[id]] = button;
+	}
+
+	fastWriteJson(root, os);
+}
+
+static touch_gui_button_id button_name_to_id(const std::string &name)
+{
+	for (u8 i = 0; i < touch_gui_button_id_END; i++) {
+		if (name == button_names[i])
+			return (touch_gui_button_id)i;
+	}
+	return touch_gui_button_id_END;
+}
+
+void ButtonLayout::deserializeJson(std::istream &is)
+{
+	layout.clear();
+
+	Json::Value root;
+	is >> root;
+
+	if (!root["layout"].isObject())
+		throw Json::RuntimeError("invalid type for layout");
+
+	Json::Value &obj = root["layout"];
+	Json::ValueIterator iter;
+	for (iter = obj.begin(); iter != obj.end(); iter++) {
+		touch_gui_button_id id = button_name_to_id(iter.name());
+		if (!isButtonAllowed(id))
+			throw Json::RuntimeError("invalid button name");
+
+		Json::Value &value = *iter;
+		if (!value.isObject())
+			throw Json::RuntimeError("invalid type for button metadata");
+
+		ButtonMeta meta;
+
+		if (!value["position_x"].isNumeric() || !value["position_y"].isNumeric())
+			throw Json::RuntimeError("invalid type for position_x or position_y in button metadata");
+		meta.position.X = value["position_x"].asFloat();
+		meta.position.Y = value["position_y"].asFloat();
+
+		if (!value["offset_x"].isNumeric() || !value["offset_y"].isNumeric())
+			throw Json::RuntimeError("invalid type for offset_x or offset_y in button metadata");
+		meta.offset.X = value["offset_x"].asFloat();
+		meta.offset.Y = value["offset_y"].asFloat();
+
+		layout.emplace(id, meta);
+	}
+}
+
+void layout_button_grid(v2u32 screensize, ISimpleTextureSource *tsrc,
+		const std::vector<touch_gui_button_id> &buttons,
+		// pos refers to the center of the button
+		const std::function<void(touch_gui_button_id btn, v2s32 pos, core::recti rect)> &callback)
+{
+	s32 cols = 4;
+	s32 rows = 3;
+	f32 screen_aspect = (f32)screensize.X / (f32)screensize.Y;
+	while ((s32)buttons.size() > cols * rows) {
+		f32 aspect = (f32)cols / (f32)rows;
+		if (aspect > screen_aspect)
+			rows++;
+		else
+			cols++;
+	}
+
+	s32 button_size = ButtonLayout::getButtonSize(screensize);
+	v2s32 spacing(screensize.X / (cols + 1), screensize.Y / (rows + 1));
+	v2s32 pos(spacing);
+
+	for (touch_gui_button_id btn : buttons) {
+		v2u32 orig_size = ButtonLayout::getTexture(btn, tsrc)->getOriginalSize();
+		v2s32 size((f32)orig_size.X / (f32)orig_size.Y * button_size, button_size);
```

Comment:
Trivial: Superfluous use of floating point - unless you'd like to add 0.5f for rounding.

```suggestion
		v2s32 size((button_size * orig_size.X) / orig_size.Y, button_size);
```

+ the same in `ButtonLayout::getRect`.